consensus/src/threshold_simplex/actors/voter/actor.rs

```diff
@@ -15,273 +13,201 @@ use commonware_cryptography::{
+    // need to track multiple proposals here.
+    finalizes: Vec<Finalize<V, D>>,
     finalization: Option<Finalization<V, D>>,
     broadcast_finalize: bool,
     broadcast_finalization: bool,
+
+    recover_latency: histogram::Timed<E>,
 }
 
 impl<
+        E: Clock,
         C: Scheme,
         V: Variant,
         D: Digest,
-        R: Reporter<Activity = Activity<V, D>>,
         S: ThresholdSupervisor<
             Seed = V::Signature,
             Index = View,
             Share = group::Share,
             PublicKey = C::PublicKey,
+            Identity = V::Public,
         >,
-    > Round<C, V, D, R, S>
+    > Round<E, C, V, D, S>
 {
-    pub fn new(clock: &impl Clock, reporter: R, supervisor: S, view: View) -> Self {
+    pub fn new(
+        context: &E,
+        supervisor: S,
+        recover_latency: histogram::Timed<E>,
+        view: View,
+    ) -> Self {
         let participants = supervisor.participants(view).unwrap().len();
+        let quorum = quorum(participants as u32);
         Self {
-            start: clock.current(),
-            reporter,
+            start: context.current(),
             supervisor,
 
             view,
-            participants,
+            quorum,
 
             leader: None,
+
+            requested_proposal_build: false,
+            requested_proposal_verify: false,
+            verified_proposal: false,
+
+            proposal: None,
+
             leader_deadline: None,
             advance_deadline: None,
             nullify_retry: None,
 
-            requested_proposal: false,
-            proposal: None,
-            verified_proposal: false,
-
-            notarized_proposals: HashMap::new(),
-            notarizes: vec![None; participants],
+            notarizes: Vec::with_capacity(participants),
             notarization: None,
             broadcast_notarize: false,
             broadcast_notarization: false,
 
-            nullifies: HashMap::new(),
+            nullifies: Vec::with_capacity(participants),
             nullification: None,
             broadcast_nullify: false,
             broadcast_nullification: false,
 
-            finalized_proposals: HashMap::new(),
-            finalizes: vec![None; participants],
+            finalizes: Vec::with_capacity(participants),
             finalization: None,
             broadcast_finalize: false,
             broadcast_finalization: false,
+
+            recover_latency,
         }
     }
 
     pub fn set_leader(&mut self, seed: V::Signature) {
         let leader = ThresholdSupervisor::leader(&self.supervisor, self.view, seed).unwrap();
-        self.leader = Some(leader);
+        let leader_index = self.supervisor.is_participant(self.view, &leader).unwrap();
+        self.leader = Some((leader, leader_index));
     }
 
-    fn add_verified_proposal(&mut self, proposal: Proposal<D>) {
+    fn add_recovered_proposal(&mut self, proposal: Proposal<D>) {
```

Comment:
We explicitly distinguish between a verified and recovered proposal (we'll never vote on a proposal we haven't verified).